ironfish-rust/src/spending.rs

```diff
@@ -62,132 +44,138 @@ pub struct SpendParams {
     /// incorporated into the proof, but is supplied to help miners verify the
     /// root hash at the time of spend.
     pub(crate) tree_size: u32,
-
-    /// Bytes to be placed into the nullifier set to verify whether this
-    /// note has been previously spent.
-    pub(crate) nullifier: Nullifier,
+    pub(crate) witness_position: u64,
+    pub(crate) auth_path: Vec<Option<(Scalar, bool)>>,
 }
 
-impl SpendParams {
-    /// Construct a new SpendParams attempting to spend a note at a given location
-    /// in the merkle tree.
+impl SpendBuilder {
+    /// Create a new [`SpendBuilder`] attempting to spend a note at a given
+    /// location in the merkle tree.
     ///
     /// This is the only time this API thinks about the merkle tree. The witness
     /// contains the root-hash at the time the witness was created and the path
     /// to verify the location of that note in the tree.
-    pub fn new(
-        spender_key: SaplingKey,
-        note: &Note,
-        witness: &dyn WitnessTrait,
-    ) -> Result<SpendParams, IronfishError> {
-        // This is a sanity check; it would be caught in proving the circuit anyway,
-        // but this gives us more information in the event of a failure
-        if !witness.verify(&MerkleNoteHash::new(note.commitment_point())) {
```

Comment:
I want to specifically call this out - this doesn't exist in the refactor (yet). Re-adding it does one of two things:
- We add it in the constructor, which requires `add_spend` to return a `Result`, and is extra computation during the setup instead of in the build phase
- We add it to the build function which will require keeping track of the witness object passed in from JS, which may lead to needing awkward lifetimes or something. 

Curious to hear others thoughts on this one. Still thinking about it.